### PR TITLE
[Doc] Add MKL-DNN operator list

### PR DESCRIPTION
**Pull request by TaoLv**
## Description ##
Add MKL-DNN optimized operator list and graph fusion pattern list.

This PR is part of the plan in https://github.com/apache/incubator-mxnet/issues/14399

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by TaoLv** [comment]
@pengzhao-intel please review~

File: docs/tutorials/mkldnn/operator_list.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+| :--: | :--: | :--: |
+| Convolution + Activation(ReLU) | MXNET_SUBGRAPH_BACKEND  | MXNET_DISABLE_MKLDNN_FUSE_CONV_RELU |
+| Convolution + elemwise_add | MXNET_SUBGRAPH_BACKEND  | MXNET_DISABLE_MKLDNN_FUSE_CONV_SUM |
+| Convolution + BatchNorm | MXNET_SUBGRAPH_BACKEND  | MXNET_DISABLE_MKLDNN_FUSE_CONV_BN |
+| Convolution + Activation(ReLu) + elemwise_add | MXNET_SUBGRAPH_BACKEND  |   |
+| Convolution + BatchNorm + Activation(ReLu) + elemwise_add | MXNET_SUBGRAPH_BACKEND  |   |
+| FullyConnected + Activation(ReLU) | MXNET_SUBGRAPH_BACKEND  | MXNET_DISABLE_MKLDNN_FUSE_FC_RELU |
+| Convolution (INT8) + re-quantization | MXNET_SUBGRAPH_BACKEND  |   |
+| FullyConnected (INT8) + re-quantization | MXNET_SUBGRAPH_BACKEND  |   |
+
```
Comment (pengzhao-intel) [inline docs/tutorials/mkldnn/operator_list.md]: The `Enable` column is repeated, I think we can remove this column and only leave the `Disabe` column

File: docs/tutorials/mkldnn/operator_list.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+To try these features out, you can install MXNet MKL-DNN backend through pip:
```
Comment (pengzhao-intel) [inline docs/tutorials/mkldnn/operator_list.md]: the install info is already in the below readme page, so I think we can remove in here.

File: docs/tutorials/mkldnn/operator_list.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+<!--- Unless required by applicable law or agreed to in writing, -->
+<!--- software distributed under the License is distributed on an -->
+<!--- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY -->
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+# MKL-DNN Operator list
+
+MXNet MKL-DNN backend provides optimized implementations for various opertors covering a broad range of applications including image classification, object detection, natural language processing. We also provide the lower precision version for part of these operators on CPU leveraging the DL Boost technology from Intel. On computation graph level, a set of graph fusion pass and quantization pass is implemneted based on the sugraph feature of MXNet. To help users understanding MKL-DNN backend better, the tables below summarize the list of supported operators, data types and functionalities. As the community keeps working on more new features for MKL-DNN backend, the tables will be updated continuously.
```
Comment (pengzhao-intel) [inline docs/tutorials/mkldnn/operator_list.md]: Make the description as simple as possible?

**Comment by TaoLv** [comment]
@xinyu-intel @ciyongch please help to review. Thank you.

File: docs/install/index.md [hunk 4]
```diff
@@ -187,7 +187,12 @@ $ pip install mxnet --pre
 
 </div> <!-- End of master-->
 <hr> <!-- pip footer -->
-MXNet offers MKL pip packages that will be much faster when running on Intel hardware.
+MXNet offers MKL pip packages that will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in [performance on Intel CPU](https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu).
```
Comment (xinyu-intel) [inline docs/install/index.md]: change hyper-link format according to L196

File: docs/tutorials/mkldnn/operator_list.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+| Operator           | Function                   | FP32 Training (backward) | FP32 Inference | INT8 Inference |
```
Comment (xinyu-intel) [inline docs/tutorials/mkldnn/operator_list.md]: maybe not support markdown table, use text format table instead.

File: docs/tutorials/mkldnn/operator_list.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+export MXNET_SUBGRAPH_BACKEND=MKLDNN
+```
+
+And disable `Convolution + Activation(ReLU)` fusion by:
+
+```
+export MXNET_DISABLE_MKLDNN_FUSE_CONV_RELU=1
+```
+
+| Fusion pattern                                            | Enable                  | Disable                             |
```
Comment (xinyu-intel) [inline docs/tutorials/mkldnn/operator_list.md]: something wrong with table render.

File: docs/tutorials/mkldnn/MKLDNN_README.md [hunk 7]
```diff
@@ -19,7 +19,9 @@
 
 A better training and inference performance is expected to be achieved on Intel-Architecture CPUs with MXNet built with [Intel MKL-DNN](https://github.com/intel/mkl-dnn) on multiple operating system, including Linux, Windows and MacOS.
 In the following sections, you will find build instructions for MXNet with Intel MKL-DNN on Linux, MacOS and Windows.
-
+
+Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](http://mxnet.incubator.apache.org/tutorials/mkldnn/operator_list.html)
```
Comment (xinyu-intel) [inline docs/tutorials/mkldnn/MKLDNN_README.md]: period

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by TaoLv** [comment]
Hi @aaronmarkham May I have your help? CI complains about converting markdown table:
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwebsite/detail/PR-14891/3/pipeline/85#step-123-log-1335

File: docs/install/index.md [hunk 8]
```diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
 $ pip install mxnet
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
```
Comment (aaronmarkham) [inline docs/install/index.md]:
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../..//faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

File: docs/install/index.md [hunk 9]
```diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
...
+```
+
 </div> <!-- End of v1-4-0 -->
 <div class="v1-3-1">
 
 ```
 $ pip install mxnet==1.3.1
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
```
Comment (aaronmarkham) [inline docs/install/index.md]:
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

File: docs/install/index.md [hunk 10]
```diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
...
+```
+
 </div> <!-- End of v1-3-1 -->
 <div class="v1-2-1">
 
 ```
 $ pip install mxnet==1.2.1
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
```
Comment (aaronmarkham) [inline docs/install/index.md]:
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

File: docs/install/index.md [hunk 11]
```diff
@@ -185,9 +203,15 @@ $ pip install mxnet==0.11.0
 $ pip install mxnet --pre
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
```
Comment (aaronmarkham) [inline docs/install/index.md]:
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

File: docs/tutorials/mkldnn/MKLDNN_README.md [hunk 12]
```diff
@@ -19,7 +19,9 @@
 
 A better training and inference performance is expected to be achieved on Intel-Architecture CPUs with MXNet built with [Intel MKL-DNN](https://github.com/intel/mkl-dnn) on multiple operating system, including Linux, Windows and MacOS.
 In the following sections, you will find build instructions for MXNet with Intel MKL-DNN on Linux, MacOS and Windows.
-
+
+Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](http://mxnet.incubator.apache.org/tutorials/mkldnn/operator_list.html).
```
Comment (aaronmarkham) [inline docs/tutorials/mkldnn/MKLDNN_README.md]:
```suggestion
Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](../mkldnn/operator_list.md).
```

**Comment by TaoLv** [comment]
Thank you for the great suggestions, @aaronmarkham .

File: docs/tutorials/mkldnn/operator_list.md (hunk 1 above)
Comment (TaoLv) [reply docs/tutorials/mkldnn/operator_list.md]: Fixed.

File: docs/tutorials/mkldnn/operator_list.md (hunk 2 above)
Comment (TaoLv) [reply docs/tutorials/mkldnn/operator_list.md]: Yes, removed.

File: docs/tutorials/mkldnn/operator_list.md (hunk 3 above)
Comment (TaoLv) [reply docs/tutorials/mkldnn/operator_list.md]: Changed according to @aaronmarkham 's suggestion.

File: docs/install/index.md (hunk 4 above)
Comment (TaoLv) [reply docs/install/index.md]: Fixed.

File: docs/tutorials/mkldnn/MKLDNN_README.md (hunk 7 above)
Comment (TaoLv) [reply docs/tutorials/mkldnn/MKLDNN_README.md]: Fixed.

File: docs/tutorials/mkldnn/operator_list.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+| **Quantization**   | 1D-4D input                | N                        | N              | Y              |
+| **Dequantization** | 1D-4D input                | N                        | N              | Y              |
+| **Requantization** | 1D-4D input                | N                        | N              | Y              |
+
+Besides direct operator optimizations, we also provide graph fusion passes listed in the table below. Users can choose to enable or disable these fusion patterns through environmental variables.
+
+For example, you can enable all fusion passes by:
+
+```
+export MXNET_SUBGRAPH_BACKEND=MKLDNN
```
Comment (ciyongch) [inline docs/tutorials/mkldnn/operator_list.md]: `all fusion passes` might not be accurate enough here.
`MKLDNN` backend is only applicable to the FP32 mode fusion, while `MKLDNN_POST_QUANTIZE` is applicable to the INT8 mode fusion (quantized_op +requantize/dequantize).

Comment (TaoLv) [reply docs/tutorials/mkldnn/operator_list.md]: Nice to know! How many subgraph backend options do we have now? What about INT8 convolution + relu fusion?

File: docs/tutorials/mkldnn/operator_list.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+And disable `Convolution + Activation(ReLU)` fusion by:
+
+```
+export MXNET_DISABLE_MKLDNN_FUSE_CONV_RELU=1
+```
```
Comment (ciyongch) [inline docs/tutorials/mkldnn/operator_list.md]: There's also other options for end-use to choose, such as:
`MXNET_DISABLE_MKLDNN_CONV_OPT` to disable all MKLDNN convolution optimization pass.
`MXNET_DISABLE_MKLDNN_FC_OPT` to disable all MKLDNN FullyConnected optimization pass.

File: docs/tutorials/mkldnn/operator_list.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+| :--:                                                      | :--:                                |
+| Convolution + Activation(ReLU)                            | MXNET_DISABLE_MKLDNN_FUSE_CONV_RELU |
+| Convolution + elemwise_add                                | MXNET_DISABLE_MKLDNN_FUSE_CONV_SUM  |
+| Convolution + BatchNorm                                   | MXNET_DISABLE_MKLDNN_FUSE_CONV_BN   |
+| Convolution + Activation(ReLu) + elemwise_add             |                                     |
+| Convolution + BatchNorm + Activation(ReLu) + elemwise_add |                                     |
+| FullyConnected + Activation(ReLU)                         | MXNET_DISABLE_MKLDNN_FUSE_FC_RELU   |
+| Convolution (INT8) + re-quantization                      |                                     |
+| FullyConnected (INT8) + re-quantization                   |                                     |
+
```
Comment (ciyongch) [inline docs/tutorials/mkldnn/operator_list.md]: miss a pattern here: `FullyConnected (INT8) + re-quantization + de-quantization`

File: docs/tutorials/mkldnn/operator_list.md (hunk 13 above)
Comment (ciyongch) [reply docs/tutorials/mkldnn/operator_list.md]: Currently, there're only two options as mentioned above.
There's no fusion pass for `INT8 conv +relu`, the current flow is doing the fusion of `FP32 conv +relu` firstly, and then quantizing `FP32 sg_conv` to `INT8 sg_conv`.

Comment (aaronmarkham) [reply docs/tutorials/mkldnn/operator_list.md]: > Thank you for the information. I think we need to put them into this page: https://mxnet.incubator.apache.org/versions/master/faq/env_var.html

Let's maintain them in one spot and then reference where appropriate. Either way some mention is needed in the env_var page.

**Comment by TaoLv** [comment]
@aaronmarkham @ciyongch @pengzhao-intel please take another look - I added some description about how the subgraph backend variable is used and added a reference link in the env var page.

File: docs/faq/env_var.md [hunk 16]
```diff
@@ -280,6 +280,11 @@ When USE_PROFILER is enabled in Makefile or CMake, the following environments ca
   - Values: Int ```(default=4)```
   - This variable controls how many CuDNN dropout state resources to create for each GPU context for use in operator.
 
+* MXNET_SUBGRAPH_BACKEND
+  - Values: String ```(default="")```
+  - This variable controls the subgraph partitioning in MXNet.
+  - This variable is used to perform MKL-DNN FP32 operator fusion and quantization. Please refer to [MKL-DNN operator list](../tutorials/mkldnn/operator_list.md) for how this variable is used and the list of fusion pass.
```
Comment (aaronmarkham) [inline docs/faq/env_var.md]:
```suggestion
  - This variable is used to perform MKL-DNN FP32 operator fusion and quantization. Please refer to the [MKL-DNN operator list](../tutorials/mkldnn/operator_list.md) for how this variable is used and the list of fusion passes.
```

File: docs/install/index.md [hunk 17]
```diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
 $ pip install mxnet
 ```
 
+MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../..//faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```
Comment (aaronmarkham) [inline docs/install/index.md]: extra slash for a reason? that might have been my mistake...
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

File: docs/tutorials/mkldnn/MKLDNN_README.md [hunk 18]
```diff
@@ -1,25 +1,27 @@
...
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
 # Build/Install MXNet with MKL-DNN
 
 A better training and inference performance is expected to be achieved on Intel-Architecture CPUs with MXNet built with [Intel MKL-DNN](https://github.com/intel/mkl-dnn) on multiple operating system, including Linux, Windows and MacOS.
 In the following sections, you will find build instructions for MXNet with Intel MKL-DNN on Linux, MacOS and Windows.
 
+Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](../mkldnn/operator_list.md).
```
Comment (aaronmarkham) [inline docs/tutorials/mkldnn/MKLDNN_README.md]:
```suggestion
Please find MKL-DNN optimized operators and other features in the [MKL-DNN operator list](../mkldnn/operator_list.md).
```

**Comment by TaoLv** [comment]
Thank you for the fix @aaronmarkham . I'm not familiar with the compilation from md files to html files, but I not notice there is already an index.md in the folder (and other folders under tutorials). Are you suggesting to overwrite the existing index.md?

**Comment by aaronmarkham** [comment]
> Thank you for the fix @aaronmarkham . I'm not familiar with the compilation from md files to html files, but I not notice there is already an index.md in the folder (and other folders under tutorials). Are you suggesting to overwrite the existing index.md?

Since others are working on that file right now in another PR, let's leave it alone for now.

**Comment by pengzhao-intel** [comment]
@TaoLv I merged two MKLDNN related PRs just now and please check if the doc needs the related change.
I think we can merge this PR when all is finalized.

**Comment by TaoLv** [comment]
> @TaoLv I merged two MKLDNN related PRs just now and please check if the doc needs the related change.
> I think we can merge this PR when all is finalized.

Sure. Will change accordingly~

**Comment by TaoLv** [comment]
@pengzhao-intel  Change the environmental variable name from `MKLDNN_POST_QUANTIZE` to `MKLDNN_QUANTIZE` accordingly. And also remove the `ReLU` constraint since we support more activation fusion now. Please review again~

**Comment by pengzhao-intel** [comment]
It's great @TaoLv, please retrigger the CI and I will merge soon.

**Comment by pengzhao-intel** [comment]
update `env_var.md` again :(

File: docs/install/index.md (hunk 8 above)
Comment (TaoLv) [reply docs/install/index.md]: Hi @aaronmarkham, the link here will be compiled to https://mxnet.incubator.apache.org/versions/faq/perf.md#intel-cpu which is a blank page. Do you think I can use http://mxnet.io/faq/perf.html#intel-cpu instead here?

Comment (TaoLv) [reply docs/install/index.md]: @pengzhao-intel

**Pull request merged**